## Exploring eBay Car Sales Data

This data is originally scraped form a german car sales eBay website by kaggle user 'orgesleka'

Data dictionary for the dataset:

* dateCrawled - When the ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing.
* price - The listed selling price of the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The type of vehicle.
* yearOfRegistration - The year in which the car was first registered.
* gearbox - The type of transmission.
* powerPS - The power of the car in PS.
* model - The car model name.
* odometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
autos = pd.read_csv("autos.csv" , encoding = 'latin1')

In [ ]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [ ]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [ ]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

* we will change the columns in autos dataframe from camelcase format to snakecase format and change some of the names to better understanding


In [ ]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [ ]:
autos.rename(columns={'yearOfRegistration':'registration_year',
                      'monthOfRegistration':'registration_month',
                      'notRepairedDamage':'unrepaired_damage',
                      'dateCreated':'ad_created'}, inplace=True)

In [ ]:
import re
def camel_to_snake(camelString):
    return re.sub(r'([A-Z]+)',r'_\1',camelString).lower()

In [ ]:
for i,column in enumerate(autos.columns):
    autos.columns.values[i] = camel_to_snake(column)


In [ ]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [ ]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Data Exploration and Cleaning

In [ ]:
autos.describe(include='all') # with include='all' we get both categorical and numeric columns

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Our initial observations:
* `seller` column and `offer_type` column has where nearly all of the values same.
* `nr_of_pictures` column looks odd, we will investigate this further.


In [ ]:
autos['nr_of_pictures'].value_counts().sort_index(ascending=True)

0    50000
Name: nr_of_pictures, dtype: int64

As we see above `nr_of_pictures` column has `0` for every row. So we can drop this column.

Since `seller` and `offer_type` columns has the same situation like `nr_of_pictures` column we will drop these columns too.

Reorganizing `price` column:
* Price column contains dollar sign and comma. To change this column type from text to numeric dtype we need to remove the dollar sign and replace the comma with dot

In [ ]:
autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(int)

/tmp/ipykernel_64/3544741403.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  autos['price']=autos['price'].str.replace('$','').str.replace(',','').astype(int)


In [ ]:
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [ ]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [ ]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [ ]:
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','').astype(int)

In [ ]:
autos['odometer']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: int64

In [ ]:
autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [ ]:
autos.head(0)

Empty DataFrame
Columns: [date_crawled, name, seller, offer_type, price, abtest, vehicle_type, registration_year, gearbox, power_ps, model, odometer_km, registration_month, fuel_type, brand, unrepaired_damage, ad_created, nr_of_pictures, postal_code, last_seen]
Index: []

In [ ]:
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [ ]:
autos['offer_type'].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In [ ]:
autos['price'].unique().shape

(2357,)

In [ ]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [ ]:
autos['price'].value_counts().sort_index(ascending=True)

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64

As we look at the price values of cars in this site we see there is 1421 record for 0 dollar. Also there is a very high numbers at the high end of the sorting. These numbers might be outliers let's look at these values closer

In [ ]:
print(autos['price'].value_counts().sort_index(ascending=True).head(30))
print(autos['price'].value_counts().sort_index(ascending=True).tail(20))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
55       2
59       1
60       9
65       5
66       1
Name: price, dtype: int64
197000      1
198000      1
220000      1
250000      1
259000      1
265000      1
295000      1
299000      1
345000      1
350000      1
999990      1
999999      2
1234566     1
1300000     1
3890000     1
10000000    1
11111111    2
12345678    3
27322222    1
99999999    1
Name: price, dtype: int64


As we see there is record's for the price under the 50 dollars. These numbers are suspicious for a car price so we will clean it from df.

When we look at the high end of the records we can see there is a high jump after 350000, so this might also be outlier. These values also will be cleaned from df.

In [ ]:
autos = autos.drop(autos[((autos['price']<50) | (autos['price']>350000))].index)

In [ ]:
autos['price'].value_counts().sort_index(ascending=True)

50        49
55         2
59         1
60         9
65         5
          ..
265000     1
295000     1
299000     1
345000     1
350000     1
Name: price, Length: 2323, dtype: int64

In [ ]:
autos['odometer_km'].describe()

count     48347.000000
mean     125875.752373
std       39618.501091
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [ ]:
autos['odometer_km'].value_counts().sort_index(ascending=True)

5000        782
10000       249
20000       759
30000       778
40000       815
50000      1011
60000      1153
70000      1215
80000      1414
90000      1733
100000     2105
125000     5045
150000    31288
Name: odometer_km, dtype: int64

We can say that the km values here are rounded.

## Exploring the Date Columns

In [ ]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

There are a number of columns with date information:

- `date_crawled`
- `registration_month`
- `registration_year`
- `ad_created`
- `last_seen`

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.


In [ ]:
autos['date_crawled']

0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
2        2016-03-26 18:57:24
3        2016-03-12 16:58:10
4        2016-04-01 14:38:50
                ...         
49995    2016-03-27 14:38:19
49996    2016-03-28 10:50:25
49997    2016-04-02 14:44:48
49998    2016-03-08 19:25:42
49999    2016-03-14 00:42:12
Name: date_crawled, Length: 48347, dtype: object

In [ ]:
(autos['date_crawled'].iloc[0:10].value_counts(normalize=True,dropna=False).sort_index())

2016-03-12 16:58:10    0.1
2016-03-16 13:47:02    0.1
2016-03-16 18:55:19    0.1
2016-03-20 17:55:21    0.1
2016-03-21 13:47:45    0.1
2016-03-22 16:51:34    0.1
2016-03-26 17:47:46    0.1
2016-03-26 18:57:24    0.1
2016-04-01 14:38:50    0.1
2016-04-04 13:38:56    0.1
Name: date_crawled, dtype: float64

In [ ]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        ) #sort index method rank date in ascending order(earliest to latest)

2016-03-05    0.025358
2016-03-06    0.014065
2016-03-07    0.036031
2016-03-08    0.033260
2016-03-09    0.033053
2016-03-10    0.032225
2016-03-11    0.032556
2016-03-12    0.036983
2016-03-13    0.015658
2016-03-14    0.036672
2016-03-15    0.034294
2016-03-16    0.029495
2016-03-17    0.031584
2016-03-18    0.012907
2016-03-19    0.034749
2016-03-20    0.037789
2016-03-21    0.037293
2016-03-22    0.032949
2016-03-23    0.032267
2016-03-24    0.029433
2016-03-25    0.031501
2016-03-26    0.032267
2016-03-27    0.031088
2016-03-28    0.034894
2016-03-29    0.034108
2016-03-30    0.033715
2016-03-31    0.031832
2016-04-01    0.033735
2016-04-02    0.035535
2016-04-03    0.038575
2016-04-04    0.036528
2016-04-05    0.013051
2016-04-06    0.003165
2016-04-07    0.001386
Name: date_crawled, dtype: float64

In [ ]:
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
        )

2016-04-07    0.001386
2016-04-06    0.003165
2016-03-18    0.012907
2016-04-05    0.013051
2016-03-06    0.014065
2016-03-13    0.015658
2016-03-05    0.025358
2016-03-24    0.029433
2016-03-16    0.029495
2016-03-27    0.031088
2016-03-25    0.031501
2016-03-17    0.031584
2016-03-31    0.031832
2016-03-10    0.032225
2016-03-23    0.032267
2016-03-26    0.032267
2016-03-11    0.032556
2016-03-22    0.032949
2016-03-09    0.033053
2016-03-08    0.033260
2016-03-30    0.033715
2016-04-01    0.033735
2016-03-29    0.034108
2016-03-15    0.034294
2016-03-19    0.034749
2016-03-28    0.034894
2016-04-02    0.035535
2016-03-07    0.036031
2016-04-04    0.036528
2016-03-14    0.036672
2016-03-12    0.036983
2016-03-21    0.037293
2016-03-20    0.037789
2016-04-03    0.038575
Name: date_crawled, dtype: float64

In [ ]:
(autos['ad_created']
    .str[0:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
    )

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038824
2016-04-04    0.036900
2016-04-05    0.011769
2016-04-06    0.003247
2016-04-07    0.001241
Name: ad_created, Length: 76, dtype: float64

In [ ]:
(autos['last_seen']
    .str[0:10]
    .value_counts(normalize=True, dropna=False)
    .sort_index()
    )

2016-03-05    0.001076
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007343
2016-03-09    0.009639
2016-03-10    0.010611
2016-03-11    0.012390
2016-03-12    0.023745
2016-03-13    0.008873
2016-03-14    0.012638
2016-03-15    0.015885
2016-03-16    0.016464
2016-03-17    0.028089
2016-03-18    0.007322
2016-03-19    0.015802
2016-03-20    0.020622
2016-03-21    0.020601
2016-03-22    0.021408
2016-03-23    0.018595
2016-03-24    0.019774
2016-03-25    0.019174
2016-03-26    0.016775
2016-03-27    0.015575
2016-03-28    0.020849
2016-03-29    0.022318
2016-03-30    0.024738
2016-03-31    0.023828
2016-04-01    0.022876
2016-04-02    0.024883
2016-04-03    0.025172
2016-04-04    0.024510
2016-04-05    0.124972
2016-04-06    0.221627
2016-04-07    0.132087
Name: last_seen, dtype: float64

In [ ]:
autos['registration_year'].describe()

count    48347.000000
mean      2004.718845
std         87.786604
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Acceptable `registration_year`:

The describe func for `registration_year` columns returns us min value as 1000 and max value for 9999 which is can't acceptable. Because of car's inventory does not align with this timestamp. Acceptable timestamp for car registration must be in between 1900's and 2016(which is the year that this website scraped)

## Dealing with Incorrect Registariton Year Data

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [ ]:
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0] # Returns the percentage of the data out of the range of 1900-2016 to all data

0.038761453657931204

Given that this is less than 4% of our data, we will remove these rows.

In [ ]:
#autos = autos.drop(autos[autos['registration_year']<1900 & autos['registration_year']>2016])

In [ ]:
autos = autos[autos['registration_year'].between(1900,2016)]

In [ ]:
autos['registration_year'].value_counts(normalize=True).head(10)

2000    0.067136
2005    0.062854
1999    0.062058
2004    0.058077
2003    0.057991
2006    0.057453
2001    0.056635
2002    0.053321
1998    0.050589
2007    0.048932
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

## Exploring Brand Column

In [ ]:
unique_vehicle_brands = autos['brand'].unique()
unique_vehicle_brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [ ]:
autos['brand'].value_counts(normalize=True)

volkswagen        0.211327
bmw               0.109956
opel              0.107310
mercedes_benz     0.096508
audi              0.086609
ford              0.069976
renault           0.047189
peugeot           0.029910
fiat              0.025714
seat              0.018247
skoda             0.016418
nissan            0.015321
mazda             0.015256
smart             0.014180
citroen           0.014051
toyota            0.012760
hyundai           0.010049
sonstige_autos    0.009618
volvo             0.009102
mini              0.008779
mitsubishi        0.008177
honda             0.007854
kia               0.007101
alfa_romeo        0.006649
porsche           0.006047
suzuki            0.005939
chevrolet         0.005702
chrysler          0.003529
dacia             0.002647
daihatsu          0.002518
jeep              0.002281
land_rover        0.002109
subaru            0.002109
saab              0.001657
jaguar            0.001571
daewoo            0.001485
trabant           0.001377
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings.  Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [ ]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts>0.05].index
common_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [ ]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos['brand']==brand]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 5424,
 'bmw': 8376,
 'opel': 2996,
 'mercedes_benz': 8663,
 'audi': 9373,
 'ford': 3762}

Of the top 5 brands, there is a distinct price gap:

* Audi, Mercedes and BMW are most expensive
* Ford and Opel less expensive.
* Volkswagen is in between - this may explain its popularity, it may have the vehicles in each price ends.

### Exploring Mileage

For the top 6 brands, let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

* it's difficult to compare more than two aggregate series objects if we want to extend to more columns
* we can't compare more than a few rows from each series object
* we can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we can combine the data from both series objects into a single dataframe (with a shared index) and display the dataframe directly. To do this, we'll need to learn two pandas methods:

pandas series constructor

pandas dataframe constructor

In [ ]:
bpm_series = pd.Series(brand_mean_prices) # with this series method the keys in the brand_mean_prices dictionary became the index in the series object.We can create a single-column dataframe from this series object
print(bpm_series)

volkswagen       5424
bmw              8376
opel             2996
mercedes_benz    8663
audi             9373
ford             3762
dtype: int64


In [ ]:
type(bpm_series)

pandas.core.series.Series

In [ ]:
bpm_series.index

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [ ]:
df= pd.DataFrame(bpm_series, columns=['mean_price']) # adding column parameter to name the column, otherwise it will be named as default 0
df

mean_price
volkswagen           5424
bmw                  8376
opel                 2996
mercedes_benz        8663
audi                 9373
ford                 3762

In [ ]:
brands_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos['brand']==brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brands_mean_mileage[brand] = int(mean_mileage)

brands_mean_mileage

{'volkswagen': 128771,
 'bmw': 132651,
 'opel': 129440,
 'mercedes_benz': 130923,
 'audi': 129254,
 'ford': 124295}

In [ ]:
mean_mileages = pd.Series(brands_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)
print(mean_mileages)
print(mean_prices)

bmw              132651
mercedes_benz    130923
opel             129440
audi             129254
volkswagen       128771
ford             124295
dtype: int64
audi             9373
mercedes_benz    8663
bmw              8376
volkswagen       5424
ford             3762
opel             2996
dtype: int64


In [ ]:
#Creating new dataframe for comparing mean mileages and mean prices together
mean_values_for_brands = pd.DataFrame(mean_mileages,columns=['mean_mileage'])
mean_values_for_brands

mean_mileage
bmw                  132651
mercedes_benz        130923
opel                 129440
audi                 129254
volkswagen           128771
ford                 124295

In [ ]:
mean_values_for_brands['mean_price'] = mean_prices
mean_values_for_brands

mean_mileage  mean_price
bmw                  132651        8376
mercedes_benz        130923        8663
opel                 129440        2996
audi                 129254        9373
volkswagen           128771        5424
ford                 124295        3762

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands.  There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.